// electron-compile/lib/read-only-compiler.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _electronCompileCache = require('electron-compile-cache');

var _electronCompileCache2 = _interopRequireDefault(_electronCompileCache);

var ReadOnlyCompiler = (function (_CompileCache) {
  _inherits(ReadOnlyCompiler, _CompileCache);

  function ReadOnlyCompiler(compilerInformation, mimeType) {
    _classCallCheck(this, ReadOnlyCompiler);

    _get(Object.getPrototypeOf(ReadOnlyCompiler.prototype), 'constructor', this).call(this);

    this.compilerInformation = compilerInformation;
    this.mimeType = mimeType;
  }

  _createClass(ReadOnlyCompiler, [{
    key: 'getCompilerInformation',
    value: function getCompilerInformation() {
      return this.compilerInformation;
    }
  }, {
    key: 'compile',
    value: function compile(sourceCode, filePath, cachePath) {
      throw new Error('Asked to compile ' + filePath + ' in production!');
    }
  }, {
    key: 'getMimeType',
    value: function getMimeType() {
      return this.mimeType;
    }
  }, {
    key: 'initializeCompiler',
    value: function initializeCompiler() {
      return this.compilerInformation.version;
    }
  }], [{
    key: 'getExtensions',
    value: function getExtensions() {
      return [];
    }
  }]);

  return ReadOnlyCompiler;
})(_electronCompileCache2['default']);

exports['default'] = ReadOnlyCompiler;
module.exports = exports['default'];
